Book query

amymhaddad opened the pull request:
@paulghaddad Here's what's included in this PR:

**app.py** --` @app.route("/books", methods=["GET", "POST"])`
-Paginate the results (so only 50 results, of the 100, show up on the books page, and the user can choose to see the next 50 results)

-Sort books by title

-Filter books by publication date

**import_books.py**
-Wrote a new Book script to update my database with 100 books and book attributes

**books.html**
-Updated html to accommodate the sort, filter, and pagination features

amymhaddad committed:
Books page allows user to sort, filter, and paginate book results.

paulghaddad commented on alembic/env.py:
```diff
@@ -22,9 +22,8 @@
-# target_metadata = None
-
-from database import Base
+# target_metadata = None)
```
Get rid of this comment

paulghaddad commented on alembic/versions/d88a20bc318e_adding_publication_year_to_books_table.py:
```diff
@@ -0,0 +1,28 @@
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
In my migrations, I tend to delete these four auto-generated comments by Alembic.

paulghaddad commented on alembic/versions/d88a20bc318e_adding_publication_year_to_books_table.py:
```diff
@@ -0,0 +1,28 @@
+"""Adding publication year to books table
```
Nice migration message...these are important to understand what the migration does!

paulghaddad commented on alembic/versions/df34d3f0af24_added_a_genre_column_to_book.py:
```diff
@@ -0,0 +1,28 @@
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Ditto on comments

paulghaddad commented on app.py:
```diff
@@ -1,5 +1,8 @@
-from flask import Flask, render_template, url_for
-from upcase_booklist_app.data.books import books
+from flask import Flask, render_template, url_for, request
```
I don't believe `url_for` is used in this file

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,40 @@ def index():
 
 
-@app.route("/books")
+@app.route("/books", methods=["GET", "POST"])
```
👍

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,40 @@ def index():
+    elif request.method == "GET" and request.args:
+        # paginate the results
+        if request.args.get("page"):
+            books = db_session.query(Book).offset(50)
```
Nice job on the offset! However, I believe you'll need to use `LIMIT` to return on 10 books per page. You'll need to figure out a way to use OFFSET and LIMIT combined to return books 1-10 or 30-40, etc. This will require use to use the `?page` query param you're passing in but not using.

paulghaddad replied:
set `request.args.get("page")` to a variable and use it. Don't hardcode 50.

amymhaddad replied:
@paulghaddad Do I have this right:
```

elif request.method == "GET" and request.args:
        # paginate the results
        if request.args.get("page"):
            limit_books = request.args.get("page")
            books = db_session.query(Book).limit(10).offset(limit_books) 
```

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,40 @@ def index():
+            books = db_session.query(Book).order_by(sort_by).all()
+
+            # sort by publication date
+            books = db_session.query(Book).order_by(Book.publication_year.desc()).all()
```
I don't think you need this. Line 36 should be sufficient to sort by whatever `sort_by` query param you pass in.

paulghaddad replied:
Get rid of the query b.c the html is mapped to the query param

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,40 @@ def index():
+            books = db_session.query(Book).order_by(Book.publication_year.desc()).all()
+
+    else:
+        books = db_session.query(Book).limit(50)
```
I'd limit the results per page to 10 everywhere in this function.

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,40 @@ def index():
     """Show details for a specific book."""
 
-    return render_template("book.html", book=books[book_id])
+    books = db_session.query(Book)
```
This variable should really be singular `book`, likewise on line 53.

paulghaddad replied:
See my comment on: 

https://github.com/amymhaddad/upcase_booklist_app/pull/10/files#r339361104

paulghaddad commented on models/book.py:
```diff
@@ -1,13 +1,15 @@
 
-    id = Column(Integer(), primary_key=True)
+    id = Column(Integer, primary_key=True)
     title = Column(String(), nullable=False)
```
One thing to consider is whether to declare an attribute `NOT NULLABLE` or not. It's really a matter of whether an attribute is essential for an object. In this case, I believe `title` is. To get this in your database, you just have to change `nullable=True` and run alembic autogenerate. Gives you another chance to practice migrations.

Look at the `NOT NULL` and `NULL` entries on the Postgres documentation:

https://www.postgresql.org/docs/10/sql-createtable.html

paulghaddad replied:
I got mixed up on the above comment, but I think you'll want `NULLABLE` for `image_url` and `summary`.

paulghaddad commented on scripts/import_books.py:
```diff
@@ -1,38 +1,38 @@
+]
+
+
+books = []
```
Really nice job in this script! Looks great! :)

paulghaddad commented on templates/book.html:
```diff
@@ -16,20 +16,25 @@
-                    </div>
-                </div>
+    <div>
+        {% for book in books %}
```
Instead of passing all the `books` back to this template in your view function, you can have your query on https://github.com/amymhaddad/upcase_booklist_app/pull/10/files#diff-3f41e546893dc64b71aaacad12cad815R51

return a single book by using `filter` with `one`. See the documentation here:

https://docs.sqlalchemy.org/en/14/orm/tutorial.html#returning-lists-and-scalars

Then you won't need to do a loop and apply a filter here: `{% if book.id == book_id %}`

paulghaddad commented on templates/book.html:
```diff
@@ -16,20 +16,25 @@
+            <div class="single-book-box">
+            <div class="single-book-display">     
+
+                <img src="{{ book.image_url }}">
```
This all looks good! 💯

paulghaddad commented on templates/books.html:
```diff
@@ -18,25 +18,55 @@
+                            <option value="Advanced">Advanced</option>
+                        </select>
+                    <button class="filter-button">Filter</button>
+                    </form>
```
Indent to match its sibling on 23.

paulghaddad commented on templates/books.html:
```diff
@@ -18,25 +18,55 @@
+                        </div>
+                
+                        </form>
+                    </div>
```
Indenting looks off.

paulghaddad commented on templates/books.html:
```diff
@@ -18,25 +18,55 @@
+    
 
+            <div> 
+                {% for book in books %}
```
This template looks good!

paulghaddad commented on templates/books.html:
```diff
@@ -18,25 +18,55 @@
+            <div class="sort-books">
+                <div class="current">
+                    <p>1
+                        <a href="books?page=2">Next</a>
```
Try just adding the first three pages:

```
<a href="books?page=1">Page 1</a>
<a href="books?page=2">Page 2</a>
<a href="books?page=3">Page 3</a>
```

paulghaddad reviewed (changes requested):
Looks really good. Just a few pointers to clean up and you should be able to merge.

amymhaddad commented:
@paulghaddad Great comments!

Here's a summary of my changes:

In app.py:
-Updated `@app.route("/books/<int:book_id>")` function to return all rows for only one book

-Updated `@app.route("/books", methods=["GET", "POST"])` to paginate 10 results per page
```
if request.args.get("page"):
            limit_books = request.args.get("page")
            books = db_session.query(Book).limit(10).offset(limit_books)
```

In books.html and book.html:
-Updated spacing

In import_books.py:
-Made image_url and summary nullable

paulghaddad commented on alembic/versions/b78cdaa6e6ec_get_books_table_added_to_database.py:
```diff
@@ -10,27 +10,24 @@
+    op.create_table(
+        "books",
+        sa.Column("id", sa.Integer(), nullable=False),
+        sa.Column("author_id", sa.Integer(), nullable=True),
```
Good choices on making this and the next attribute nullable.

paulghaddad commented on alembic/versions/7b49481bf438_add_books_table.py:
```diff
@@ -10,19 +10,15 @@
 
```
Not sure what happened with this migration but there's nothing in it

paulghaddad commented on alembic/versions/c7a2d0f9b089_adding_books_to_book_table.py:
```diff
@@ -10,19 +10,15 @@
 
 
 # revision identifiers, used by Alembic.
-revision = 'c7a2d0f9b089'
```
This migration doesn't contain anything.

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,35 @@ def index():
+
+    if request.method == "POST":
+        user_genre = request.form["genre"]
+        books = db_session.query(Book).filter(Book.genre == user_genre)
```
Excellent

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,35 @@ def index():
+    elif request.method == "GET" and request.args:
+        if request.args.get("page"):
+            limit_books = request.args.get("page")
+            books = db_session.query(Book).limit(10).offset(limit_books)
```
Excellent! 👏

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,35 @@ def index():
+            sort_by = request.args.get("sort")
+            books = db_session.query(Book).order_by(sort_by).all()
+    else:
+        books = db_session.query(Book).limit(10)
```
Excellent! 💯

paulghaddad commented on app.py:
```diff
@@ -14,17 +17,35 @@ def index():
     """Show details for a specific book."""
 
-    return render_template("book.html", book=books[book_id])
+    book = db_session.query(Book).filter(Book.id == book_id).one()
```
Perfect! Nice work! 💯

paulghaddad reviewed (approved):
Really nice work P! Proud of you! 🥇
